case-lib/pipeline.sh

```diff
@@ -1,7 +1,27 @@
 #!/bin/bash
 
+# This function will evaluate pipeline parameter related shell code generated by
+# - sof-tplgreader.py (for SOF, pipeline parameters dumped from topology)
+# - sof-dump-status.py (for legacy HDA, pipeline paramters dumped from proc)
+# Args: $1: SOF topology path
+#       $2: Pipeline filter string
```

Comment:
So @plbossart  was right, you need a topology path even when `is_sof_used_` is false? Which one?

I thought this function took different arguments depending on `is_sof_used`

Reply:
@marc-hb On legacy HDA, if you manually run test case, no TPLG is required. But CI will give a fake topology, as you can see from any report, this is how CI run a test case: `TPLG=sof-hda-generic-4ch-kwd.tplg ~/sof-test/test-case/verify-tplg-binary.sh`. Some compromise with sof-framework here. Here is how func_pipeline_export called in test case: ` func_pipeline_export $tplg "type:capture" `. so the filter string may be $1 (tplg is empty if manual run) or $2 (CI always give tplg), that's the reason for `"$@"["$#"]`

Reply:
> func_pipeline_export $tplg "type:capture". so the filter string may be $1 (tplg is empty if manual run) or $2 (CI always give tplg

- This is _really_ missing from the function header!
- That's yet another demonstration that almost every parameter expansion must be quoted. Maybe it would be a too big change but with `func_pipeline_export "$tplg" "type:capture"` then `$1` would be `""` when `$tplg` is missing, the function would get a constant number of arguments and the filter would always be in `$2`. 

> that's the reason for "$@"["$#"]

... which doesn't seem to do what I understand you want. How was this tested!?

Reply:
I have run the test case on a legacy HDA platform, and an SOF platform with TPLG, both passed!

Reply:
Maybe a broken `filter_str` does not matter?

Reply:
sof-tplgreader.py and sof-dump-status.py will help to handle the filter_str, so we don't need a handler here.

Reply:
I meant: maybe a broken `filter_str` did not matter in your testing? So `filter_str` was never tested?

We digressing again and spent way too many comments and time on this. Trying to focus back on the code issues:

- `"$@"["$#"]` does not do anything anyone may want, it just doesn't work for any known definition of "work".
- `"$@"["$#"]` was not tested. Every new lines of code must be tested at least once
- The (failed) intent doesn't seem to match the function header which says nothing either about all the other, discarded arguments in the middle.
- Getting the last function argument is a bit complicated in shell and best avoided. 
- Getting the last argument is probably not a good idea in general, is it really needed here? Why not just pass an empty string first argument instead of none?

Before the next force-push what is your plan to address all these?

Reply:
@marc-hb I am going to drop the `"$@"["$#"]` thing, and add quote in every test case to `$tplg`, thus make sure the filter_str is always $2. So we don't need to think about how to get the last function argument in shell.

Reply:
> add quote in every test case to $tplg

Sounds good. You probably want to merge that first and separately to avoid a "Big Bang Integration" (the opposite of _Continuous_ Integration)

Reply:
I find that "fake topologies" sound bad too. On the other hand, an empty, **_dummy_ HDA topology file** would be OK if needed to keep sof-test APIs consistent or for some transition period. Such a dummy HDA topology should make any test FAIL FAST if the test tries using it by mistake.

Reply:
@plbossart @marc-hb Agreed. I think this requires some change in the CI framework, I will check.